onmt/dynamic/transforms/bart.py

```diff
@@ -0,0 +1,372 @@
+            "Error when insert noise."
+        return [tok for tok in result]
+
+    def rolling_noise(self, tokens, p=1.0):
+        if np.random.random() >= p:
+            return tokens
+        offset = np.random.randint(0, max(1, len(tokens) - 1) + 1)
+        return tokens[offset:] + tokens[0:offset]
+
+    def apply(self, tokens):
+        if self.vocab is None:
+            raise ValueError("Inject BART noise require a valid vocabulary.")
+
+        if self.permute_sent_ratio > 0.0:
+            tokens = self.permute_sentences(tokens, self.permute_sent_ratio)
+
+        if self.mask_ratio > 0.0:
+            tokens = self.whole_word_mask(tokens, self.mask_ratio)
+
+        if self.insert_ratio > 0.0:
+            tokens = self.insertion_noise(tokens, self.insert_ratio)
+
+        if self.rotate_ratio > 0.0:
+            tokens = self.rolling_noise(tokens, self.rotate_ratio)
+        return tokens
+
+    def __repr__(self):
+        cls_name = type(self).__name__
+        kwargs = {}
+        if self.permute_sent_ratio > 0.0:
+            kwargs['permute_sent_ratio'] = self.permute_sent_ratio
+            kwargs['full_stop_token'] = self.full_stop_token
+        if self.insert_ratio > 0.0:
+            kwargs['insert_ratio'] = self.insert_ratio
+        if self.rotate_ratio > 0.0:
+            kwargs['rotate_ratio'] = self.rotate_ratio
+        if self.random_ratio > 0.0:
+            kwargs['random_ratio'] = self.random_ratio
+        if self.mask_ratio > 0.0:
+            kwargs['mask_ratio'] = self.mask_ratio
+            kwargs['mask_length'] = self.mask_length
+            kwargs['poisson_lambda'] = self.poisson_lambda
+            kwargs['replace_length'] = self.replace_length
+        cls_args = ', '.join(
+            [f'{kw}={arg}' for kw, arg in kwargs.items()])
+        return '{}({})'.format(cls_name, cls_args)
+
+
+@register_transform(name='bart')
+class BARTNoiseTransform(Transform):
+    def __init__(self, opts):
+        super().__init__(opts)
+
+    @classmethod
+    def add_options(cls, parser):
+        """Avalilable options relate to BART."""
+        group = parser.add_argument_group('Transform/BART')
+        group.add('--permute_sent_ratio', '-permute_sent_ratio',
+                  type=float, default=0.0,
+                  help="shuffle this proportion of sentences in all inputs")
+        group.add('--rotate_ratio', '-rotate_ratio', type=float, default=0.5,
+                  help="rotate this proportion of inputs")
+        group.add('--insert_ratio', '-insert_ratio', type=float, default=0.0,
+                  help="insert this percentage of additional random tokens")
+        group.add('--random_ratio', '-random_ratio', type=float, default=0.0,
+                  help="instead of using [MASK], use random token this often")
+
+        group.add('--mask_ratio', '-mask_ratio', type=float, default=0.0,
+                  help="fraction of words/subwords that will be masked")
```

Comment:
The function using this value is called whole_word_mask. Isn't it the fraction of 'words" and not "subwords" then?

Reply:
Nevermind, just saw the following flag...

Reply:
`whole_word_mask` is the original name in fairseq. And the granularity is controlled by `mask_length`. It can be word or subword if `-mask_length` set to subword.